database/memory/db.go

```diff
@@ -535,34 +611,199 @@ func (db *MemoryDB) removeAllIndices(address types.Address) error {
 	return nil
 }
 
+func (db *MemoryDB) GetERC20EntryAtBlock(contract types.Address, holder types.Address, block uint64) (*ERC20TokenHolder, error) {
+	var tmpItem *ERC20TokenHolder
+	found := false
+	for i, item := range db.erc20BalancesDB {
+		if item.BlockNumber <= block && item.Contract == contract && item.Holder == holder {
+			if !found {
+				tmpItem = &db.erc20BalancesDB[i]
+				found = true
+			} else {
+				if item.BlockNumber > tmpItem.BlockNumber {
+					tmpItem = &db.erc20BalancesDB[i]
+				}
+			}
+		}
+	}
+	if !found {
+		return nil, database.ErrNotFound
+	}
+	return tmpItem, nil
+}
+
 func (db *MemoryDB) RecordNewERC20Balance(contract types.Address, holder types.Address, block uint64, amount *big.Int) error {
+	existingTokenEntry, errExisting := db.GetERC20EntryAtBlock(contract, holder, block-1)
+	if errExisting != nil && errExisting != database.ErrNotFound {
+		return errExisting
+	}
+
+	//add new entry
+	tokenInfo := ERC20TokenHolder{
+		Contract:    contract,
+		Holder:      holder,
+		BlockNumber: block,
+		Amount:      amount.String(),
+	}
+	db.erc20BalancesDB = append(db.erc20BalancesDB, tokenInfo)
+	/////
+	if errExisting == database.ErrNotFound {
+		return nil
+	}
+	blk := block - 1
+	existingTokenEntry.HeldUntil = &blk
+
 	return nil
 }
 
 func (db *MemoryDB) GetERC20Balance(contract types.Address, holder types.Address, options *types.TokenQueryOptions) (map[uint64]*big.Int, error) {
-	return nil, database.ErrNotImplemented
+	balanceMap := make(map[uint64]*big.Int)
+	frmBlkNum := options.BeginBlockNumber.Uint64()
+	endBlkNum := options.EndBlockNumber.Int64()
+	for _, b := range db.erc20BalancesDB {
+		if contract == b.Contract && holder == b.Holder && b.BlockNumber >= frmBlkNum && (b.BlockNumber <= uint64(endBlkNum) || endBlkNum == -1) {
+			tokAmt, success := new(big.Int).SetString(b.Amount, 10)
+			if !success {
+				return nil, errors.New("could not parse token value")
+			}
+			balanceMap[b.BlockNumber] = tokAmt
+		}
+	}
+	return balanceMap, nil
 }
 
 func (db *MemoryDB) GetAllTokenHolders(contract types.Address, block uint64, options *types.TokenQueryOptions) ([]types.Address, error) {
-	return nil, database.ErrNotImplemented
+	var holderMap = make(map[types.Address]bool)
+	for _, k := range db.erc20BalancesDB {
+		if k.Contract == contract && k.BlockNumber <= block && k.Holder != "0000000000000000000000000000000000000000" {
+			holderMap[k.Holder] = true
+		}
+	}
+	var holderArr []types.Address
+	for holdr := range holderMap {
+		holderArr = append(holderArr, holdr)
+	}
+	return holderArr, nil
 }
 
 func (db *MemoryDB) RecordERC721Token(contract types.Address, holder types.Address, block uint64, tokenId *big.Int) error {
+	//find old entry
+	existingTokenEntry, errExisting := db.ERC721TokenByTokenID(contract, block-1, tokenId)
+	if errExisting != nil && errExisting != database.ErrNotFound {
+		return errExisting
+	}
+
+	paddedTokenId := fmt.Sprintf("%085d", tokenId)
```

Comment:
since we have removed the logic of sorting based on these parts of the token id, remove the storage as well

Reply:
done